Nocturnalscream feature hyper creep

NocturnalScream opened the pull request:
Incorperatres changes of pr#1
fixes which were requested in pr1
added hypercreep option (roboports will creep forwards with powerpoles once all concrete is laid)
added option for adjusting amount of no work updates needed to trigger a reinit

known issues: 
- Square pattern of the power poles sometimes brakes for unknown reasons.
- After loading a save game, concreep appears to be broken - if you call init via a settings change it will start to work again
- also script_on_nth_tick doesn't seem to update its speed for all the reinit functions - they wont get any faster.

NocturnalScream committed:
Initial commit

added hypercreep (roboport creeping)
-still needs fixing with the powerpoles

NocturnalScream committed:
Hypercreep option added

fixed some issues with hypercreep and added an option to activate it.

NocturnalScream commented on concreep-refilled/control.lua:
```diff
@@ -55,6 +55,92 @@ function init()
+			hypercreep_builder(target_x_pos,target_y_pos,offset_power_poles,roboport)		 
+		end
+	end	
+end	
```
there must be a better way to do this - maybe something like a switch statement? maybe you can help

MarximusMaximus replied:
I'd do this like this:

```
    local roboport_x = roboport.position.x
    local roboport_y = roboport.position.y

    hypercreep_builder(roboport_x + logistic_diameter, roboport_y                    , offset_power_poles, roboport)
    hypercreep_builder(roboport_x - logistic_diameter, roboport_y                    , offset_power_poles, roboport)
    hypercreep_builder(roboport_x                    , roboport_y + logistic_diameter, offset_power_poles, roboport)
    hypercreep_builder(roboport_x                    , roboport_y - logistic_diameter, offset_power_poles, roboport)
    hypercreep_builder(roboport_x + logistic_diameter, roboport_y + logistic_diameter, offset_power_poles, roboport)
    hypercreep_builder(roboport_x - logistic_diameter, roboport_y - logistic_diameter, offset_power_poles, roboport)
    hypercreep_builder(roboport_x + logistic_diameter, roboport_y - logistic_diameter, offset_power_poles, roboport)
    hypercreep_builder(roboport_x - logistic_diameter, roboport_y + logistic_diameter, offset_power_poles, roboport)
```

NocturnalScream replied:
uhm - yeah dont know why i insisted on using a for loop here - i think it was for testing quickly what different patterns would look like but this definitely makes much more sense

NocturnalScream replied:
weirldy enough - your solution does not work as intended - the top left corner of the grid does not get build - cant quite see whats wrong here - maybe you should try the mod out in game as well - with editor extensions its rather nice to debug - doing more testing right now

NocturnalScream replied:
turns out - thats not your code - would have been really wondering there - because it should be identical to mine in terms of functionality - need to check what causes this behaviour

NocturnalScream committed:
Fixed redundant calls

Fixed some redundant calls for hypercreep_builder

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -64,16 +51,87 @@ function init()
+	script.on_nth_tick(settings.global["run every n updates"].value, creepers_update)
+end
+
+function hypercreep_builder(target_x_pos,target_y_pos,offset,roboport)
```
```
function hypercreep_builder(target_x_pos,target_y_pos,offset,roboport)
    debug_print_coroutine_was_called("hypercreep_builder()")
    local surface = roboport.surface
    local force = roboport.force
    local roboport_x_pos = roboport.position.x
    local roboport_y_pos = roboport.position.y
    local half_distance_x = math.floor((target_x_pos-roboport.position.x)/2)
    local half_distance_y = math.floor((target_y_pos-roboport.position.y)/2)
    if (
        surface.can_place_entity{name="entity-ghost", position={target_x_pos,target_y_pos}, inner_name=roboport.name, force=force} and
        surface.can_place_entity{name="entity-ghost", position={target_x_pos,target_y_pos+offset}, inner_name="big-electric-pole", force=force, build_check_type=defines.build_check_type.script_ghost} and
        surface.count_entities_filtered{area={{target_x_pos-10,target_y_pos-10},{target_x_pos+10,target_y_pos+10}},name="entity-ghost"} == 0 -- check whether we can place the roboport+power pole before trying
    ) then
        surface.create_entity{name="entity-ghost", position={target_x_pos,target_y_pos}, inner_name=roboport.name, force=force, expires=false}
        surface.create_entity{name="entity-ghost", position={target_x_pos,target_y_pos+offset}, inner_name="big-electric-pole", force=force, expires=false}
        if (
            target_y_pos ~= roboport_y_pos and
            surface.can_place_entity{name="entity-ghost", position={target_x_pos,target_y_pos-half_distance_y+offset}, inner_name="big-electric-pole", force=force, build_check_type=defines.build_check_type.script_ghost} and
            surface.count_entities_filtered{area={{target_x_pos-10,target_y_pos-half_distance_y+offset-10},{target_x_pos+10,target_y_pos-half_distance_y+offset+10}},name="entity-ghost"} == 0
        ) then
            surface.create_entity{name="entity-ghost", position={target_x_pos,target_y_pos-half_distance_y+offset}, inner_name="big-electric-pole", force=force, expires=false}
        end
        if (
            target_x_pos ~= roboport_x_pos and
            surface.can_place_entity{name="entity-ghost", position={target_x_pos-half_distance_x,target_y_pos+offset}, inner_name="big-electric-pole", force=force, build_check_type=defines.build_check_type.script_ghost} and
            surface.count_entities_filtered{area={{target_x_pos-half_distance_x-10,target_y_pos+offset-10},{target_x_pos-half_distance_x+10,target_y_pos+offset+10}},name="entity-ghost"} == 0
        ) then
            surface.create_entity{name="entity-ghost", position={target_x_pos-half_distance_x,target_y_pos+offset}, inner_name="big-electric-pole", force=force, expires=false}
        end
    end
end
```

NocturnalScream replied:
looking into this right now - i think i might have had some errors in there - I need to check not only if we expand horizontally or vertically but actually up down left or right.

NocturnalScream replied:
also i cannot copy paste this - it won't accept the line breaks in visual studio - leaving me with it putting everything in one line

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -609,6 +704,6 @@ script.on_event(
-script.on_nth_tick(RUN_EVERY_N_UPDATES, creepers_update)
 script.on_init(init)
 script.on_configuration_changed(is_valid_roboport_tile_names)
+script.on_event(defines.events.on_runtime_mod_setting_changed, runtime_settings_changed)
```
fixes not working when loading save game, add these lines:
```
script.on_nth_tick(nil) --unregister all nth tick handlers, so we can update the interval n
script.on_nth_tick(settings.global["run every n updates"].value, creepers_update)
```

NocturnalScream replied:
done

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -171,15 +241,15 @@ function _reinit_reentrant_level1_body()
 					surface and
 					surface.valid
```
we should check if the surface is in the list of desired surfaces here

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -144,7 +202,19 @@ function _reinit_reentrant_level1_start()
 
 	-- set up outer arrays to iterate
-	global.reinit_surfaces = array_shallowcopy(game.surfaces)
+	local listOfSurfaces = {}
```
we should not pre-filter the list of surfaces here, we should validate below during the loops (to allow player to edit the surface list in the middle of a loop)

NocturnalScream replied:
Think i have done that now - however not quite sure how to check now if we even have a single valid surface in case the user screws up the input - I guess I will find a use for the script.on_configuration_changed

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -224,13 +293,13 @@ function _reinit_reentrant_level2_body()
 					add_creeper_for_roboport(roboport)
 				end
-				-- need to manually increment
+				-- need to defines.build_check_type.script_ghostly increment
```
keep the word manually (b/c we're simulating a loop)

NocturnalScream replied:
ah damn - didnt intend to change this - was an all replace operation messing with this comment

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -242,37 +311,37 @@ function _reinit_reentrant_level2_end()
 	global.reinit_level_running = 1
 
-	-- need to manually increment outer loop as the inner loop ends
+	-- need to defines.build_check_type.script_ghostly increment outer loop as the inner loop ends
```
keep the word manually (b/c we're simulating a loop)

NocturnalScream replied:
same as above - accidental change

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -290,6 +359,13 @@ function get_creeper()
 	return creeper
 end
 
+function save_entity_creation(surface, name, position, inner_name, force, expires)
```
unused function?

NocturnalScream replied:
yes - sadly, i wanted to use this but ran into some error - I guess it wouldn't be efficient anyways, was just annoyed by the fact I always needed to check then build.

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -306,23 +382,38 @@ function creep(creeper)
-	local ghosts = surface.count_entities_filtered{area=area, name="tile-ghost", force=force}
+	debug_print("Surface: " .. roboport.surface.name) 
+
+	--[[ 	if surface.can_place_entity{name="roboport", position={target_x_pos,target_y_pos}, inner_name=roboport.name, force=force} and 
```
remove commented code?

NocturnalScream replied:
gone

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -306,23 +382,38 @@ function creep(creeper)
+		end ]]
+	
 
+	local ghosts = surface.count_entities_filtered{area=area, name="tile-ghost", force=force}
```
add this after this line:
```
    debug_print("# of Ghosts: " .. #ghosts)
```

NocturnalScream replied:
done

NocturnalScream replied:
mmh -  doesn't it actually would need to be just `debug_print("# of Ghosts: " .. ghosts)` as ghosts is just an int

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -306,23 +382,38 @@ function creep(creeper)
 	)
+	local available_worker_count = math.max(0, roboport.logistic_network.available_construction_robots - settings.global["minimum robot"].value)
+	if roboport.logistic_network.available_construction_robots <= settings.global["minimum robot"].value then
+		debug_print("can NOT place " .. " - minimum robots")
```
unnecessary string concatenation?

NocturnalScream replied:
ah yes - wanted to get the amount of robots in there but forgot to do that in the middle of it - guess it serves the purpose without the number anyways. fixed

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -571,7 +663,7 @@ function on_built_entity_handler(event)
 
 	local entity = event.created_entity or event.destination or event.entity
-	if not global.creepers then
+	if global.creepers=={} then
```
this can crash; change back.

NocturnalScream replied:
i dont quite see how though? init() as well as reinit() will always make sure that global.creepers exist - changed back.

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -599,6 +689,11 @@ function is_valid_roboport_tile_names()
 
+function runtime_settings_changed()
+debug_print_function_was_called("runtime_settings_changed()")
+	init()
```
instead of call `init` here, we can now just do:
```
script.on_nth_tick(nil) --unregister all nth tick handlers, so we can update the interval n
script.on_nth_tick(settings.global["run every n updates"].value, creepers_update)
```
as everything else gets read from the settings directly... unless we really really want to "restart" the creeping process from scratch if any setting is changed (such that the user immediately sees the affect of their change, rather than it happening organically later)?

MarximusMaximus commented on concreep-refilled/locale/en/locale.cfg:
```diff
@@ -10,6 +10,7 @@ cover landfill=Cover landfill
 debug=Various debug output (very spammy)
 debug function calls=Function call debug output (very spammy)
 debug coroutine calls=Coroutine call debug output (very spammy)
+hypercreep=enables hypercreep mode (world eater)
```
i kind of wanna called this "paved paradise"

NocturnalScream replied:
agreed

NocturnalScream replied:
what about changing the mod name to paved paradise? :D kidding here however it would make it more clear that this fork of conreep will have a major functionality upgrade

MarximusMaximus reviewed (changes requested):
>    Square pattern of the power poles sometimes brakes for unknown reasons.

Unsure. Might be due to already built roboports not aligning to the desired grid?

>    After loading a save game, concreep appears to be broken - if you call init via a settings change it will start to work again

Fixed in comments. Please apply.

>    also script_on_nth_tick doesn't seem to update its speed for all the reinit functions - they wont get any faster.

I think the fix in the comments fixes this too? Not totally sure. Which specific setting(s)?

NocturnalScream committed:
Refinerey of hypercreep builder

- Fixes which were requested in the review (compare to comments)
- added a new method of checking whether we already built an entity on said positions to prevent double ghosts
- cleaned up hypercreep_builder
- added surface clearing routine for hypercreep
- added charting for hypercreeping

NocturnalScream committed:
Delete control - Kopie.lua

accidentally commited a copied file

NocturnalScream committed:
Changed readability for github

Changed from tabs to whitespaces, should make it more readable on github

NocturnalScream commented:
Do you have any more additions except for bugfixes for this PR? i am thinking whether we should add auto placement of radars in order to keep everything charted inside of the hypercreep mode

NocturnalScream committed:
Remove unused Variables

- some variables were never used - removed them
- changed roboport and power pole count minimum to adjust for worst case scenario

NocturnalScream committed:
cleaned up and optimized hypercreep_builder

NocturnalScream committed:
git being weird

Dont know exactly why it forces me to commit this but well - i have to - was using a local debugger

NocturnalScream committed:
Refactoring of hypercreep builder

- overhauled the way how we find power poles and check whether they are existing already
- added the ability to clear obstacles
- added the ability to build on water (place a ghost on water and build landfill under it)
fixed some localization errors.
- preparing to enable user to chose a specific power pole

NocturnalScream commented:
known issues:
- For some reason, even though we check for it, sometimes it happens that two ghosts of the roboport + the powerpole 
  underneath are built ( happens when the roboport must be placed on water - but only if there are two roboports adjacent to it.
- We need to implement some sort of lookahead check whether the roboport which needs to be build already would touch another logistic cell - if so we should not creep - this might screw with a factory otherwise when there are roboports outside of the hypercreep grid.
- need to check whether the hypercreep call is in the correct place
- BIG ONE: once we are finished laying all concrete - that's it, it doesn't look like concreep is able to reactivate for any of the old roboports even if you change the copncreep range - i.e.e you start with 30% and want to increase to 60% - all old ones wont be touched, staying at 30%

MarximusMaximus commented:
i think radars + a lamp (or two) inside hypercreep would be good. i would even consider a gun + meteor cannon too, lol

MarximusMaximus committed:
Merge branch 'main' into pr/NocturnalScream/3

# Conflicts:
#	.gitignore
#	.vscode/settings.json

MarximusMaximus committed:
renames, alignments, ignored some warnings

MarximusMaximus committed:
cleanup build scripts, use own mods folder to run
